packages/realm-server/prerender/manager-app.ts

```diff
@@ -0,0 +1,391 @@
+  lastSeenAt: number;
+};
+
+type Registry = {
+  servers: Map<string, ServerInfo>; // key: serverUrl
+  realms: Map<string, string[]>; // realm -> array of serverUrls (deque semantics)
+  lastAccessByRealm: Map<string, number>;
+};
+
+const log = logger('prerender-manager');
+
+function now() {
+  return Date.now();
+}
+
+function normalizeURL(u: string): string {
+  // ensure no trailing slash
+  try {
+    let url = new URL(u);
+    url.hash = '';
+    url.search = '';
+    return url.toString().replace(/\/$/, '');
+  } catch (e) {
+    return u.replace(/\/$/, '');
+  }
+}
+
+async function ping(url: string, timeoutMs: number): Promise<boolean> {
+  try {
+    const ac = new AbortController();
+    const t = setTimeout(() => ac.abort(), timeoutMs).unref?.();
+    const res = await fetch(url, { method: 'GET', signal: ac.signal });
+    clearTimeout(t as any);
+    return res.ok;
+  } catch (e) {
+    return false;
+  }
+}
+
+function clientURLFromContext(ctxt: Koa.Context): string | null {
+  let headerUrl = ctxt.get('X-Prerender-Server-Url');
+  if (headerUrl) {
+    return normalizeURL(headerUrl);
+  }
+  let ip = ctxt.req.socket.remoteAddress || '';
+  if (!ip) {
+    return null;
+  }
+  // strip IPv6 prefix if present
+  if (ip.startsWith('::ffff:')) ip = ip.slice(7);
+  const defaultPort = Number(process.env.PRERENDER_SERVER_DEFAULT_PORT ?? 4223);
+  const scheme = 'http';
+  return `${scheme}://${ip}:${defaultPort}`;
+}
+
+export function buildPrerenderManagerApp(): {
+  app: Koa<Koa.DefaultState, Koa.Context>;
+  registry: Registry;
+  sweepServers: () => Promise<void>;
+} {
+  const app = new Koa<Koa.DefaultState, Koa.Context>();
+  const router = new Router();
+  const registry: Registry = {
+    servers: new Map(),
+    realms: new Map(),
+    lastAccessByRealm: new Map(),
+  };
+
+  const multiplex = Math.max(1, Number(process.env.PRERENDER_MULTIPLEX ?? 1));
+  const proxyTimeoutMs = Math.max(
+    1000,
+    Number(process.env.PRERENDER_SERVER_TIMEOUT_MS ?? 30000),
+  );
+  const healthcheckTimeoutMs = Math.max(
+    100,
+    Number(process.env.PRERENDER_HEALTHCHECK_TIMEOUT_MS ?? 1000),
+  );
+
+  // health
+  router.head('/', async (ctxt) => {
+    ctxt.status = 200;
+  });
+  router.get('/', async (ctxt) => {
+    ctxt.set('Content-Type', 'application/json');
+    ctxt.body = JSON.stringify({ ready: true });
+    ctxt.status = 200;
+  });
+
+  // register server
+  router.post('/prerender-servers', async (ctxt) => {
+    try {
+      let req = await fetchRequestFromContext(ctxt);
+      let raw = await req.text();
+      let requestBody: any = {};
+      if (raw) {
+        try {
+          requestBody = JSON.parse(raw);
+        } catch (e) {
+          log.debug('Invalid JSON body on registration; treating as empty:', e);
+        }
+      }
+
+      let attrs = requestBody?.data?.attributes || {};
+      let capacity: number = Number(attrs.capacity ?? 4);
+      let url: string | undefined = attrs.url;
+      if (!url) {
```

Comment:
Is auth needed for this or the other endpoints that make changes?

Reply:
No auth necessary, the callers (our workers) will all be in the same VPC as the server--this is not exposed to the outside world, so we can just use a security group to make sure there no unintended access.